### DISCUSSION TITLE
[0.22] How to fix `*mut pyo3::Python<'static> cannot be shared between threads safely` when using `with_gil` and `allow_threads`

### DISCUSSION THREAD

**Discussion by relf** [PyO3/pyo3#4622] [Questions] [answered]
I have implemented an optimizer in Rust and use it from Python. I pass a Python function and use `pyo3-log` to get some info during the optimization process. With the previous API, it works fine, I got something like:

```rust
   fn minimize(&self, py: Python, fun: PyObject) -> PyResult<OptimResult> {
        let fun = fun.to_object(py);
        let obj = move |x: &ArrayView2<f64>| -> Array2<f64> {
            Python::with_gil(|py| {
                let args = (x.to_owned().into_pyarray_bound(py),);
                let res = fun.call1(py, args).unwrap();
                let pyarray: &PyArray2<f64> = res.extract(py).unwrap();
                pyarray.to_owned_array()
            })
        };
        let optimizer = builder::optimize(obj).build();

        let res = py.allow_threads(|| {
            optimizer.run().expect("optimize the objective function")
        });
        ...
```

now I am trying to upgrade to the new API (PyO3 0.22), I've tried with the following code but I get a compilation error :

```rust    
fn minimize(&self, py: Python, fun: PyObject) -> PyResult<OptimResult> {
  let fun = fun.bind(py);
  let obj = move |x: &ArrayView2<f64>| -> Array2<f64> {
      Python::with_gil(|py| {
          let args = (x.to_owned().into_pyarray_bound(py),);
          let res = fun.call1(args).unwrap();
          let pyarray = res.downcast_into::<PyArray2<f64>>().unwrap();
          pyarray.to_owned_array()
      })
  };
  let optimizer= builder::optimize(obj).build();

  let res = py.allow_threads(|| {  #  Compilation error `*mut pyo3::Python<'static>` cannot be shared between threads safely...
      optimizer.run().expect("optimize the objective function")
  });
  ...
```
I've seen there are some developments in progress around `allow_threads` in [#3646](https://github.com/PyO3/pyo3/pull/3646) and some discussions https://github.com/PyO3/pyo3/discussions/3788 but I do not see how to fix it. Thanks for your help and this awesome project!

**Comment by mejrs** [comment]
`fun` has a Python token, so the closure that captures does too. (This is also explained in the compiler error).

**Reply by relf** [reply]
Ok so do you know how should I write this to make it work? I mean the old version works like a charm what is the translation using the new API?

**Reply by mejrs** [reply]
I think you can simply delete the `let fun = fun.bind(py);` line.

**Reply by relf** [reply, marked as answer]
Thanks for your reply! I just had to put the `fun.bind(py)` call within the closure.
